Introduce identity for remote nodes

tgraf opened the pull request:
This is an initial attempt to resolve the lack of policy granularity to differentiate the local host from remote worker nodes.

The PR is complete enough as a basis for discussion and likely not fully complete yet.

Current reasoning for this approach:
 * An identity to include non-local nodes allows to continue for the cluster entity to work without much change. Thew new identity is simply added to the selector.
 * It allows to whitelist NodePort destinations by allowing "host" and "remote-node"
 * This approach does *not* allow to apply CIDR rules to worker nodes.

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/8841)
<!-- Reviewable:end -->

aanm commented on Documentation/concepts/terminology.rst:
```diff
@@ -202,36 +202,46 @@ order to allow communication to network endpoints which are not managed by
+| reserved:host           | The local host. Any traffic that originates from  |
+|                         | or is designated to one of the local host IPs.    |
++-------------------------+---------------------------------------------------+
+| reserved:non-local-node | The collection of all non-local cluster hosts.    |
```
hmmm this looks really similar to the `reserved:cluster` identity we had

tgraf replied:
Yes. It's pretty similar to what we had before we moved the cluster to be an entity to include worker nodes, host, managed pods and unmanaged pods.

jrajahalme commented on bpf/lib/policy.h:
```diff
@@ -72,7 +72,7 @@ static inline bool __inline__ inherit_identity_from_host(struct __sk_buff *skb,
 {
-	return identity < UNMANAGED_ID;
+	return identity < UNMANAGED_ID || identity == NON_LOCAL_NODE_ID;
 }
```
Update the comment above.

jrajahalme commented on pkg/defaults/defaults.go:
```diff
@@ -293,4 +293,7 @@ const (
 	K8sServiceCacheSize = 128
+
+	// EnableNonLocalNodeIdentity is the default value for option.EnableNonLocalNodeIdentity
+	EnableNonLocalNodeIdentity = false
```
Is it possible to flip this to 'true' in some later release?

tgraf replied:
It's turned on by default via the Helm default value so new deployments have this enabled by default but upgrades will preserve the existing behavior.

jrajahalme reviewed (approved):
I like this, wonder if we can default to using the non-local-node identity without requiring the new command line argument? Could the upgrade process introduce a new "legacy-mode" command line argument instead?

joestringer commented on pkg/node/manager/manager.go:
```diff
@@ -281,6 +281,11 @@ func (m *Manager) NodeUpdated(n node.Node) {
+	remoteHostIdentity := identity.ReservedIdentityHost
+	if option.Config.EnableNonLocalNodeIdentity {
+		remoteHostIdentity = identity.ReservedIdentityNonLocalNode
+	}
```
Should we also start to treat `pkg/node/addressing.NodeInternalIP` as `identity.ReservedIdentityNonLocalNode`? the code just below entirely skips propagating IPcache entries for these IP types, which includes network-facing IPs in ENI environments (amongst other cases). In practice today those IPs are only inserted if encryption is enabled, and are otherwise considered `world`.

tgraf replied:
Yes. I see that the internal IP is only added when encryption is enabled which seems wrong.

joestringer commented on Documentation/install/upgrade.rst:
```diff
@@ -322,6 +322,12 @@ New ConfigMap Options
+    for non-local cluster nodes. This allows to treat local and remote cluster
+    nodes differently from a policy perspective. The option is enabled by
+    default for new deployments when generated via Helm and disabled by default
+    for existing deployments.
```
What does it mean that it is "disabled by default for existing deployments" ? The next minor release X.Y upgrade will cause the new default to be picked up if users follow the [current upgrade guide](http://docs.cilium.io/en/latest/install/upgrade/#step-2-option-a-generate-yaml-using-helm-recommended)

joestringer replied:
Ah, I guess that rather than following the (2A) options, the user must follow the (2B) options.

joestringer reviewed (commented):
Looks good, seems like this provides a level of granularity around infrastructure IPs which could be handy in some cases. I have a couple of minor comments below.

aanm commented on pkg/identity/numericidentity.go:
```diff
@@ -75,6 +75,10 @@ const (
 	// received any labels yet.
 	ReservedIdentityInit
 
+	// ReservedIdentityNonLocalNode is the identity given to worker nodes
```
Can you be a little more specific here?

I assume "to worker nodes in any connected cluster" should be "to worker nodes, running a cilium-agent, in any connected cluster"

tgraf replied:
Not really, we also identify nodes where Cilium is not running on.

tgraf replied:
I've clarified the comment a bit

aanm commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -63,6 +63,7 @@ cilium-agent [flags]
       --enable-ipv6                                           Enable IPv6 support (default true)
       --enable-k8s-event-handover                             Enable k8s event handover to kvstore for improved scalability
       --enable-node-port                                      Enable NodePort type services by Cilium (beta)
+      --enable-non-local-node-identity                        Enable use of non-local node identity
```
Can you add in the commit message why is it important to maintain backwards compatibility? In other words, why is it important that we won't start using this identity right away by default?

tgraf replied:
Done

aanm commented on Documentation/install/upgrade.rst:
```diff
@@ -281,6 +281,19 @@ Removed options
+    for non-local cluster nodes. This allows to treat local and remote cluster
+    nodes differently from a policy perspective. The option is enabled by
+    default for new deployments when generated via Helm and disabled by default
+    for existing deployments to preserve backward compatibility.
```
Same here, please document why is it important to keep backward compatibility for this case.

tgraf replied:
Done

joestringer commented on Documentation/install/upgrade.rst:
```diff
@@ -313,6 +313,30 @@ Removed options
 
+.. _1.7_upgrade_notes:
+
+1.7 Upgrade Notes
```
nit, I made this comment in another PR too but this title is already present in the docs above these new lines.

tgraf replied:
Fixed

joestringer reviewed (changes requested):
LGTM. What's the plan for (regression) testing this functionality?

Seemingly trivial question, but I found myself wondering while reviewing this: Is there a reason the identity isn't just "remote-node"? This seems easier to understand to me. We can always get the functionality in and fix this up later though. (not sure if there's some nuance here with clusterwide vs global etc. but the current name doesn't indicate anything around that either.)

I was also thinking about how we determine the setting of this option when trying to debug clusters. I imagine that it will show up in the `cilium debuginfo` output inside a sysdump, and also the BPF header will have `NON_LOCAL_NODE_ID` configured to either `6` (true) or `1` (false). I guess that's probably enough.

Minor nit below.

jrajahalme reviewed (approved):
Looks good to me, with one minor suggestion (or question for discussion): Would "other-node" sound more natural than "non-local-node"?

joestringer commented on pkg/option/config.go:
```diff
@@ -639,6 +639,9 @@ const (
 	EnableLocalNodeRoute = "enable-local-node-route"
+
+	// EnableRemoteNodeIdentity enables use of the non-local-node identity
+	EnableRemoteNodeIdentity = "enable-non-local-node-identity"
```
I notice there's still some `non-local` references here, in the docs, and one or two other places like code comments. Did you intend to also change these?

tgraf replied:
Yes, I've already changed it. I will push after I'm done adding the CI tests.

joestringer reviewed (changes requested):
There's still a few references to `non-local-node` that need fixing up.

tgraf committed:
node: Assign remote node identity to all node IPs

Node IPs other than the Cilium internal IPs where only added to the ipcache so
far when encryption was enabled. This commit adds all node IPs to the ipcache
and assign the correct node identity.

The code is cleaned up while doing so. `n.GetNodeIP(false)` always returns the
IPv4 node address and is not dependant on the current address while iterating
over all node addresses. Calling it inside the loop is therefore unnecessary as
the result is always the same. Similarly, the `nodeIP4` variable is redundant
as the nodeIP is always an IPv4 address anyway.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
bpf: Map HOST_ID to REMOTE_NODE_ID when encapsulating

When encapsulating to other nodes, any traffic originating from the local host
must be presented as remote node traffic as it is being received. Therefore, map
any use of HOST_ID to REMOTE_NODE_ID.

Consequently, drop any traffic received as HOST_ID in bpf_overlay.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
node: Use remote-node identity when enabled

Assign the new remote-node identity to all remote hosts if the option is
enabled. The option is disabled by default as it has the potential to change
existing policy rules where users are relying on the behavior that allowing
from the host will allow traffic from all hosts in the cluster.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
policy: Add remote-node entity

This allows matching remote nodes with a policy entity.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
helm: Enable remote-node identity for all new deployments by default

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
policy: Introduce reserved identity for remote cluster nodes

Signed-off-by: Thomas Graf <thomas@cilium.io>

joestringer commented on bpf/bpf_overlay.c:
```diff
@@ -182,6 +189,10 @@ static inline int handle_ipv4(struct __sk_buff *skb, __u32 *identity)
 		*identity = key.tunnel_id;
+
+		if (*identity == HOST_ID) {
+			return DROP_INVALID_IDENTITY;
```
This is bound to drop some traffic that originates at nodes running older versions, at the point of freshly-upgraded nodes. I presume that this will be fine as users will need to opt-in to this feature during upgrade.

joestringer replied:
This concern is basically addressed in the docs already.

joestringer commented on pkg/node/manager/manager.go:
```diff
@@ -280,24 +280,22 @@ func (m *Manager) backgroundSync() {
-		isOwning := ipcache.IPIdentityCache.Upsert(address.IP.String(), nodeIP, n.EncryptionKey, nil, ipcache.Identity{
-			ID:     identity.ReservedIdentityHost,
+		isOwning := ipcache.IPIdentityCache.Upsert(address.IP.String(), tunnelIP, n.EncryptionKey, nil, ipcache.Identity{
+			ID:     remoteHostIdentity,
```
Does this path get hit for the local node as well? Is it possible we unintentionally end up updating the ipcache with local IPs pointing to ID: `identity.ReservedIdentityRemoteNode`?

tgraf replied:
I have validated that the local host entries are correct:
```
fe80::a49e:ebff:fe11:91ef/128   1 0 0.0.0.0
10.0.2.15/32                    1 0 0.0.0.0
10.16.228.47/32                 1 0 0.0.0.0
f00d::a10:0:0:9a42/128          1 0 0.0.0.0
fc00::10ca:1/128                1 0 0.0.0.0
192.168.34.11/32                1 0 0.0.0.0
192.168.33.11/32                1 0 0.0.0.0
```

Although I think you are correct that it can be called and the reason why it works is because  `syncEndpointsAndHostIPs()` is called first and will insert the local IPs with the correct identity and because it uses `source.Local`, it can't be overwritten by any subsequent updates.

tgraf replied:
I've changed the condition to:
```
if option.Config.EnableRemoteNodeIdentity && n.Source != source.Local {
```

joestringer commented on Documentation/install/upgrade.rst:
```diff
@@ -325,6 +325,21 @@ New ConfigMap Options
+    idea to enable this option as you upgrade as it improves the default
+    security posture of your cluster. If you have policy rules matching on host
+    with the intent of allow from all cluster nodes, it is recommended to
+    modify those policy rules to explicitly allow the entity ``non-local-node``
```
```suggestion
    modify those policy rules to explicitly allow the entity ``remote-node``
```

joestringer commented on pkg/option/config.go:
```diff
@@ -651,6 +651,9 @@ const (
 	// control plane, e.g. when using the managed etcd feature
 	EnableWellKnownIdentities = "enable-well-known-identities"
+
+	// EnableRemoteNodeIdentity enables use of the non-local-node identity
```
```suggestion
	// EnableRemoteNodeIdentity enables use of the remote-node identity
```

joestringer commented on pkg/option/config.go:
```diff
@@ -1292,6 +1295,9 @@ type DaemonConfig struct {
 	// control plane, e.g. when using the managed etcd feature
 	EnableWellKnownIdentities bool
+
+	// EnableRemoteNodeIdentity enables use of the non-local-node identity
```
```suggestion
	// EnableRemoteNodeIdentity enables use of the remote-node identity
```

joestringer commented on bpf/bpf_overlay.c:
```diff
@@ -72,6 +72,13 @@ static inline int handle_ipv6(struct __sk_buff *skb, __u32 *identity)
+
+		/* Any node encapsulating will map any HOST_ID source to be
+		 * presented as REMOTE_NODE_ID, therefore any attempt to signal
+		 * HOST_ID as source from a remote node can be droppped. */
```
I think we need this treatment for the `ipv6_store_flowlabel()` call in `bpf_lxc.c` as well?

tgraf replied:
I had looked at this and the only use of `ipv6_store_flowlabel()` is this:
```
        if (ipv6_store_flowlabel(skb, l3_off, SECLABEL_NB) < 0)
```

`SECLABEL_NB` is always guaranteed to be an endpoint identity. That said, adding a check for HOST_ID and dropping such attempts seems correct so I will add that.

tgraf replied:
@joestringer I've added a commit to drop such packets.

joestringer reviewed (changes requested):
Some more minor comments, but this is very close :+1:

tgraf committed:
bpf: Prohibit the IPv6 flowlabel transport to represent HOST_ID

The existing codepath was already never presenting HOST_ID as the source
identity and with the introduction of REMOTE_NODE_ID, this is guarnateed to be
impossible. Drop such packets with an appropriate error code.

Signed-off-by: Thomas Graf <thomas@cilium.io>

joestringer reviewed (approved):
Just needs rebase.
